File: godwit/src/pl/shockah/godwit/gl/color/HSVColorSpace.java
package pl.shockah.godwit.gl.color;

import javax.annotation.Nonnull;

import lombok.EqualsAndHashCode;
import pl.shockah.godwit.Math2;
import pl.shockah.godwit.fx.ease.Easing;

@EqualsAndHashCode
public class HSVColorSpace implements ColorSpace<HSVColorSpace> {
	public float h;
	public float s;
	public float v;

	public HSVColorSpace(float h, float s, float v) {
		this.h = h;
		this.s = s;
		this.v = v;
	}

	@Nonnull public static HSVColorSpace from(@Nonnull RGBColorSpace rgb) {
		float max = Math2.max(rgb.r, rgb.g, rgb.b);
		float min = Math2.min(rgb.r, rgb.g, rgb.b);
		float range = max - min;

		float h, s, v;
		if (range == 0)
			h = 0;
		else if (max == rgb.r)
			h = (60 * (rgb.g - rgb.b) / range + 360) % 360;
		else if (max == rgb.g)
			h = 60 * (rgb.b - rgb.r) / range + 120;
		else
			h = 60 * (rgb.r - rgb.g) / range + 240;

		if (max > 0)
			s = 1 - min / max;
		else
			s = 0;

		v = max;

		return new HSVColorSpace(h / 360f, s, v);
	}

	@Override
	public String toString() {
		return String.format("[HSVColorSpace: H:%.3f S:%.3f V:%.3f]", h, s, v);
	}

	@Override
	@Nonnull public RGBColorSpace toRGB() {
		float h = this.h * 360f;
		float x = (h / 60f + 6) % 6;
		int i = (int)x;
		float f = x - i;
		float p = v * (1 - s);
		float q = v * (1 - s * f);
		float t = v * (1 - s * (1 - f));
		switch (i) {
			case 0:
				return new RGBColorSpace(v, t, p);
			case 1:
				return new RGBColorSpace(q, v, p);
			case 2:
				return new RGBColorSpace(p, v, t);
			case 3:
				return new RGBColorSpace(p, q, v);
			case 4:
				return new RGBColorSpace(t, p, v);
			default:
				return new RGBColorSpace(v, p, q);
		}
	}

	@Override
	public float getDistance(@Nonnull HSVColorSpace other) {
		return (float)Math.sqrt(
				Math.pow(Math.abs(Math2.deltaAngle(h * 360f, other.h * 360f) / 360f), 2)
						+ Math.pow(s - other.s, 2)
						+ Math.pow(v - other.v, 2)
		);
	}

	@Override
	@Nonnull public HSVColorSpace ease(@Nonnull HSVColorSpace other, float f) {
		float h2 = Math2.deltaAngle(this.h, other.h) >= 0 ? other.h : other.h - 1f;
		float h = Easing.linear.ease(this.h, h2, f);
		if (h < 0)
			h += 1f;
		if (h > 0)
			h -= 1f;

		return new HSVColorSpace(
				h,
				Easing.linear.ease(s, other.s, f),
				Easing.linear.ease(v, other.v, f)
		);
	}
}

File: core/src/pl/shockah/godwit/color/HSVColorSpace.kt
package pl.shockah.godwit.color

import pl.shockah.godwit.geom.Degrees
import kotlin.math.pow
import kotlin.math.sqrt

data class HSVColorSpace(
		var h: Float,
		var s: Float,
		var v: Float
) : ColorSpace<HSVColorSpace> {
	companion object {
		fun from(rgb: RGBColorSpace): HSVColorSpace {
			val max = maxOf(rgb.r, rgb.g, rgb.b)
			val min = minOf(rgb.r, rgb.g, rgb.b)
			val range = max - min

			val h: Float
			val s: Float
			val v: Float
			h = when {
				range == 0f -> 0f
				max == rgb.r -> (60 * (rgb.g - rgb.b) / range + 360) % 360
				max == rgb.g -> 60 * (rgb.b - rgb.r) / range + 120
				else -> 60 * (rgb.r - rgb.g) / range + 240
			}

			s = if (max > 0) 1 - min / max else 0f
			v = max

			return HSVColorSpace(h / 360f, s, v)
		}
	}

	override fun copy(): HSVColorSpace = HSVColorSpace(h, s, v)

	override val rgb: RGBColorSpace
		get() {
			val h = this.h * 360f
			val x = (h / 60f + 6) % 6
			val i = x.toInt()
			val f = x - i
			val p = v * (1 - s)
			val q = v * (1 - s * f)
			val t = v * (1 - s * (1 - f))
			return when (i) {
				0 -> RGBColorSpace(v, t, p)
				1 -> RGBColorSpace(q, v, p)
				2 -> RGBColorSpace(p, v, t)
				3 -> RGBColorSpace(p, q, v)
				4 -> RGBColorSpace(t, p, v)
				else -> RGBColorSpace(v, p, q)
			}
		}

	override fun getDistance(other: HSVColorSpace): Float {
		var delta = Degrees(h * 360f) delta Degrees(other.h * 360f)
		if (delta.value < 0)
			delta = Degrees(delta.value + 360f)
		return sqrt((delta.value / 360f).pow(2) + (s - other.s).pow(2) + (v - other.v).pow(2))
	}

	override fun ease(other: HSVColorSpace, f: Float): HSVColorSpace {
		val delta = Degrees(h * 360f) delta Degrees(other.h * 360f)
		val h2 = if (delta.value >= 0) other.h else other.h - 1f
		var h = this.h.ease(h2, f)
		if (h < 0)
			h += 1f
		if (h > 0)
			h -= 1f

		return HSVColorSpace(
				h,
				s.ease(other.s, f),
				v.ease(other.v, f)
		)
	}
}